README.md
# text2speech
[![text2speech](https://github.com/kmulvey/text2speech/actions/workflows/release_build.yml/badge.svg)](https://github.com/kmulvey/text2speech/actions/workflows/release_build.yml) [![Go Report Card](https://goreportcard.com/badge/github.com/kmulvey/text2speech)](https://goreportcard.com/report/github.com/kmulvey/text2speech) [![Go Reference](https://pkg.go.dev/badge/github.com/kmulvey/imageconvert.svg)](https://pkg.go.dev/github.com/kmulvey/imageconvert)

Convert text to speech using [AWS Polly](https://aws.amazon.com/polly/).

## Dependencies
This depends on [oto](https://github.com/hajimehoshi/oto#prerequisite) which has some prerequisites.

## Examples
### Pipe text:
```
echo "The saddest aspect of life right now is that science gathers knowledge faster than society gathers wisdom." | ./text2speech -bucket your-s3-bucket
```
### File Input
`./text2speech -bucket your-s3-bucket -input text`

### MP3 output:
`./text2speech -bucket your-s3-bucket -input text -output audio.mp3  # this will only write the file, it will not play it`

### Displaying a dashboard to monitor progress
`./text2speech -bucket your-s3-bucket -input text -dashboard`

### Print help:
`./text2speech -h`


CHANGELOG.md

1.0.0
=============
2022-10-12

* report card badge (2ac28be8)
* readme notes (95d9ac4a)
* handle error (6ff7908a)
* reset max size (edcd0668)
* remove buffering (5e49fc29)
* dashboard finally working (9f11a6be)
* make dashboard optional (62457ca2)
* dashboard deps (892fa825)
* stream to dashboard (675c9f82)
* handle errors and start replacing progress bar (6307769b)
* channel (de1c0877)
* split test under polly limit (c2c6d40d)
* splitInput (af2eae7b)
* Merge pull request #52 from kmulvey/dependabot/go_modules/github.com/hajimehoshi/oto/v2-2.3.1 (3bcff97c)
* Bump github.com/hajimehoshi/oto/v2 from 2.3.0 to 2.3.1 (ed860bd0)
* Merge pull request #51 from kmulvey/dependabot/go_modules/github.com/schollz/progressbar/v3-3.11.0 (cd09d917)
* Bump github.com/schollz/progressbar/v3 from 3.10.1 to 3.11.0 (fb6e6a01)

0.7.2
=============
2022-09-06

* .goreleaser.yaml typo (e6bbaa11)

0.7.0
=============
2022-09-06

* add space after 100%, add version print (be4539f0)
* add space after 100%, add version print (0e31d3ec)

0.6.0
=============
2022-09-06

* badges (55225c78)
* deps (2c0ba6db)
* 100 (22e43dd3)
* Merge pull request #48 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2/config-1.17.5 (df35d82c)
* Bump github.com/aws/aws-sdk-go-v2/config from 1.17.1 to 1.17.5 (26aee99e)
* Merge pull request #47 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2/service/polly-1.17.7 (7ae64be5)
* Bump github.com/aws/aws-sdk-go-v2/service/polly from 1.17.4 to 1.17.7 (b85dff5c)
* Merge pull request #46 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2/service/s3-1.27.9 (be1795af)
* Bump github.com/aws/aws-sdk-go-v2/service/s3 from 1.27.5 to 1.27.9 (a9166280)
* Merge pull request #40 from kmulvey/dependabot/go_modules/github.com/hajimehoshi/oto/v2-2.3.0 (8c3cfb57)
* Bump github.com/hajimehoshi/oto/v2 from 2.2.0 to 2.3.0 (90f3d83d)
* Merge pull request #33 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2/config-1.17.1 (e3badc6e)
* Bump github.com/aws/aws-sdk-go-v2/config from 1.15.15 to 1.17.1 (c6ee4d04)
* Merge pull request #31 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2/service/s3-1.27.5 (fbbf0bdf)
* Bump github.com/aws/aws-sdk-go-v2/service/s3 from 1.27.2 to 1.27.5 (6d6aa8f6)
* Merge pull request #30 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2/service/polly-1.17.4 (0038c543)
* Bump github.com/aws/aws-sdk-go-v2/service/polly from 1.17.1 to 1.17.4 (830a2107)
* Merge pull request #29 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2-1.16.11 (f9382cab)
* Bump github.com/aws/aws-sdk-go-v2 from 1.16.8 to 1.16.11 (1f6ed8ff)
* progressbar opts (41865021)
* build badge (0f821445)
* bump deps (65eb3fba)
* Merge pull request #18 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2/service/polly-1.17.1 (101f6138)
* Bump github.com/aws/aws-sdk-go-v2/service/polly from 1.16.2 to 1.17.1 (3ced383f)
* Merge pull request #17 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2/config-1.15.15 (fd2b474e)
* Bump github.com/aws/aws-sdk-go-v2/config from 1.15.14 to 1.15.15 (7bd1e1bc)
* Merge pull request #15 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2/service/s3-1.27.2 (b2412820)
* Bump github.com/aws/aws-sdk-go-v2/service/s3 from 1.27.1 to 1.27.2 (5d3e33c6)
* lint errors (749b3081)
* switch to hajimehoshi/go-mp3 (aec8f353)
* progress bar (9c333a4f)
* dont stop (80602b1a)
* Merge pull request #13 from kmulvey/dependabot/go_modules/github.com/hajimehoshi/oto/v2-2.2.0 (3df7ae8e)
* Bump github.com/hajimehoshi/oto/v2 from 2.1.0 to 2.2.0 (0eebfb3c)
* Merge pull request #12 from kmulvey/dependabot/go_modules/github.com/sirupsen/logrus-1.9.0 (5960a445)
* ffmpeg (11821648)
* Bump github.com/sirupsen/logrus from 1.8.1 to 1.9.0 (0980e20c)
* Merge pull request #11 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2/config-1.15.14 (19619a29)
* oto (5c73d4a8)
* Bump github.com/aws/aws-sdk-go-v2/config from 1.15.13 to 1.15.14 (d35411a8)

0.5.0
=============
2022-07-09

* Merge pull request #9 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2/service/polly-1.16.2 (fcfc0610)
* Bump github.com/aws/aws-sdk-go-v2/service/polly from 1.16.1 to 1.16.2 (103386e6)
* Merge pull request #8 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2/service/s3-1.27.1 (64515d15)
* Bump github.com/aws/aws-sdk-go-v2/service/s3 from 1.26.12 to 1.27.1 (6962636b)
* Merge pull request #7 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2/config-1.15.13 (4de29471)
* Bump github.com/aws/aws-sdk-go-v2/config from 1.15.12 to 1.15.13 (2e550ce4)
* Merge pull request #3 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2/config-1.15.12 (c36ae699)
* Bump github.com/aws/aws-sdk-go-v2/config from 1.15.11 to 1.15.12 (0914d0af)
* Merge pull request #2 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2/service/s3-1.26.12 (136d8c79)
* Bump github.com/aws/aws-sdk-go-v2/service/s3 from 1.26.11 to 1.26.12 (056f500c)
* Merge pull request #5 from kmulvey/dependabot/go_modules/github.com/aws/aws-sdk-go-v2/service/polly-1.16.1 (96384515)
* Bump github.com/aws/aws-sdk-go-v2/service/polly from 1.15.3 to 1.16.1 (43011d31)

0.4.0
=============
2022-06-28

* desp (406f899d)
* max count later (44bad0ad)
* validate voiceID (ae842b64)
* me (ad744b32)
* vuln scan (4fd6190a)
* deps (c0ce5bd5)

0.3.0
=============
2022-06-20

* deps (7526bd6b)
* linting (a8f93f85)
* correct if() block (7ee43070)

0.2.0
=============
2022-06-16

* remove mac for now (20b76fee)
* more deps (eb4c2864)
* v3 (c9a8f050)
* rebuild (03699b07)
* build (ae862364)
* more build stuff (d9bae28c)
* Update README.md (1baebc37)
* build flags (76ee7c15)
* only amd64 (e638acf1)

0.1.0
=============
2022-06-16

* comments (9b5a77f9)
* ignore binary (6c6de0ab)
* oto/v2 (4b842d1d)
* update deps (868181ae)
* install deps (19563380)
* comment for now (85708a3d)
* build (d6c0e9d9)
* todo (2bc74113)
* linting (5c568195)

audio.go
package main

import (
	"context"
	"errors"
	"fmt"
	"io"
	"net/url"
	"os"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/polly"
	"github.com/aws/aws-sdk-go-v2/service/polly/types"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/hajimehoshi/go-mp3"
	"github.com/hajimehoshi/oto/v2"
)

// synthesizeText takes text and sends it to AWS polly for processing, the polly object containing the audio.
func synthesizeText(ctx context.Context, pollyClient *polly.Client, s3Client *s3.Client, logs chan string, bucket, voiceID, text string) (*s3.GetObjectOutput, string, error) {

	inputTask := &polly.StartSpeechSynthesisTaskInput{OutputFormat: "mp3", OutputS3BucketName: aws.String(bucket), Text: aws.String(text), VoiceId: types.VoiceId(voiceID)}
	task, err := pollyClient.StartSpeechSynthesisTask(ctx, inputTask)
	if err != nil {
		return nil, "", fmt.Errorf("failed to convert to speech, %w", err)
	}

	var fileURI string
	for {
		var sTask, err = pollyClient.GetSpeechSynthesisTask(ctx, &polly.GetSpeechSynthesisTaskInput{TaskId: task.SynthesisTask.TaskId})
		if err != nil {
			return nil, "", fmt.Errorf("failed to get task status, %w", err)
		}

		if sTask.SynthesisTask.TaskStatus == types.TaskStatusCompleted {
			fileURI = *sTask.SynthesisTask.OutputUri
			break
		} else if sTask.SynthesisTask.TaskStatus == types.TaskStatusFailed {
			return nil, "", fmt.Errorf("task failed: err: %w;  reason: %s", err, *sTask.SynthesisTask.TaskStatusReason)
		}

		logs <- fmt.Sprintf("Synthesis running... status: %s, id: %s \n", sTask.SynthesisTask.TaskStatus, *sTask.SynthesisTask.TaskId)

		time.Sleep(time.Second * 5)
	}

	s3File, err := url.Parse(fileURI)
	if err != nil {
		return nil, "", fmt.Errorf("failed to parse s3 uri, %w", err)
	}

	var path = strings.Split(s3File.Path, "/")
	if len(path) != 3 {
		return nil, "", fmt.Errorf("s3 path is not three elements: %d, %+v", len(path), path)
	}

	voice, err := s3Client.GetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(bucket),
		Key:    &path[2],
	})

	return voice, path[2], err
}

// play does just that (using oto)
func play(sound io.Reader) error {

	// Decode file
	var decodedMp3, err = mp3.NewDecoder(sound)
	if err != nil {
		panic("mp3.NewDecoder failed: " + err.Error())
	}

	otoCtx, ready, err := oto.NewContext(decodedMp3.SampleRate(), 2, 2)
	if err != nil {
		return fmt.Errorf("oto.NewContext: %w", err)
	}
	<-ready

	var player = otoCtx.NewPlayer(decodedMp3)
	player.Play()
	for player.IsPlaying() {
		time.Sleep(time.Millisecond)
	}

	if err := player.Close(); err != nil {
		return fmt.Errorf("error closing player: %w", err)
	}
	return nil
}

// getDuration uses ffmpeg to get the duration of the audio because its not
// returned by polly. The return value is seconds. Hopefully polly adds the
// duration in the return data and we dont have to do this anymore.
func getDuration(filename string) (int, error) {

	if _, err := os.Stat(filename); err != nil {
		return 0, err
	}

	var imagePath = EscapeFilePath(filename)
	//nolint:gosec
	out, err := exec.Command("bash", "-c", fmt.Sprintf("ffmpeg -hide_banner -i %s -f null /dev/null", imagePath)).CombinedOutput()
	if err != nil {
		return 0, fmt.Errorf("error running ffmpeg on image: %s, error: %s, output: %s", imagePath, err.Error(), out)
	}

	// create regex and find the duration in the ffmpeg output
	var durationRegex = regexp.MustCompile(`Duration:\s\d\d:\d\d:\d\d.\d\d`)
	var match = durationRegex.FindString(string(out))
	if match == "" {
		return 0, errors.New("unable to get duration from ffmpeg")
	}
	match = strings.ReplaceAll(match, "Duration: ", "")

	// iterate through the string and parse the time: e.g. (00:04:40.66)
	var durationArr = strings.Split(match, ":")
	var duration int
	for i, digits := range durationArr {
		switch i {
		case 0:
			// hours
			num, err := strconv.Atoi(digits)
			if err != nil {
				return 0, err
			}
			duration = num * 3600
		case 1:
			// minutes
			num, err := strconv.Atoi(digits)
			if err != nil {
				return 0, err
			}
			duration += num * 60
		case 2:
			// seconds
			num, err := strconv.ParseFloat(digits, 32)
			if err != nil {
				return 0, err
			}
			duration += int(num)
		default:
			return 0, errors.New("could not parse duration")
		}
	}
	return duration, nil
}


input.go
package main

import (
	"bufio"
	"context"
	"fmt"
	"io"
	"strings"
	"unicode"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

// readInput reads in chunks and returns a string
func readInput(reader io.Reader) (string, error) {

	var r = bufio.NewReader(reader)
	var buf = make([]byte, 0, 4*1024)
	var builder = strings.Builder{}
	for {
		n, err := r.Read(buf[:cap(buf)])
		buf = buf[:n]
		if n == 0 {
			if err == nil {
				continue
			}
			if err == io.EOF {
				break
			}
			return "", fmt.Errorf("input buffer error: %w", err)
		}

		builder.Write(buf)

		if err != nil && err != io.EOF {
			return "", fmt.Errorf("input buffer error: %w", err)
		}
	}

	return strings.TrimSpace(builder.String()), nil
}

// splitInput splits the text input into chunks less than MAX_CHAR_COUNT
// which is the current polly limit per job.
func splitInput(fulltext string) []string {
	var numWords = len(strings.Fields(fulltext))

	if numWords < MAX_CHAR_COUNT {
		return []string{fulltext}
	}

	var result []string
	var fulltextWords = strings.FieldsFunc(fulltext, isSpace)
	var lastEndIndex int
	for i := MAX_CHAR_COUNT; i >= 0; i-- {
		if strings.HasSuffix(fulltextWords[i], ".") {
			result = append(result, strings.Join(fulltextWords[lastEndIndex:i+1], " "))
			lastEndIndex = i + 1
			i += MAX_CHAR_COUNT
			if i >= len(fulltextWords) {
				result = append(result, strings.Join(fulltextWords[lastEndIndex:], " "))
				break
			}
		}
	}

	return result
}

func isSpace(c rune) bool {
	return unicode.IsSpace(c)
}

// deleteS3File deletes the file that polly writes to s3 after we are done playing it.
func deleteS3File(ctx context.Context, s3Client *s3.Client, bucket, key string) error {
	var _, err = s3Client.DeleteObject(ctx, &s3.DeleteObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
	})
	return err
}

// EscapeFilePath escapes spaces in the filepath used for an exec() call
func EscapeFilePath(file string) string {
	var r = strings.NewReplacer(" ", `\ `, "(", `\(`, ")", `\)`, "'", `\'`, "&", `\&`, "@", `\@`)
	return r.Replace(file)
}


dashboard.go
package main

import (
	"context"
	"fmt"
	"os"

	"github.com/mum4k/termdash"
	"github.com/mum4k/termdash/cell"
	"github.com/mum4k/termdash/container"
	"github.com/mum4k/termdash/linestyle"
	"github.com/mum4k/termdash/terminal/tcell"
	"github.com/mum4k/termdash/terminal/terminalapi"
	"github.com/mum4k/termdash/widgets/gauge"
	"github.com/mum4k/termdash/widgets/text"
	"github.com/sirupsen/logrus"
	log "github.com/sirupsen/logrus"
)

// Errorlog allows us to capture errors without getting trapped by the dashboard.
// Hopefully this can be removed when all the bugs are eradicated.
var Errorlog = logrus.New()

func init() {
	file, err := os.OpenFile("dashboard_error.log", os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
	if err == nil {
		Errorlog.Out = file
	} else {
		log.Info("Failed to log to file, using default stderr")
	}
}

// Dashboard is our terminal visualization
type Dashboard struct {
	Logs        *text.Text
	ProgressBar *gauge.Gauge
}

// NewDashboard is a constructor that builds the terminal dashboard as well as spins up two goroutines (below) that
// receive data from the two channels passed in.
func NewDashboard(ctx context.Context, cancel context.CancelFunc, playbackProgress chan PlaybackProgress, logs chan string) (*Dashboard, error) {
	var dashboard = new(Dashboard)

	var term, err = tcell.New()
	if err != nil {
		return nil, err
	}
	defer term.Close()

	dashboard.ProgressBar, err = gauge.New(
		gauge.Height(1),
		gauge.Color(cell.ColorNumber(33)),
		gauge.Border(linestyle.Light),
		gauge.BorderTitle("Section progress"),
	)
	if err != nil {
		return nil, err
	}

	dashboard.Logs, err = text.New(text.RollContent(), text.WrapAtWords())
	if err != nil {
		return nil, err
	}

	c, err := container.New(
		term,
		container.Border(linestyle.Light),
		container.BorderTitle("PRESS Q TO QUIT"),
		container.SplitHorizontal(
			container.Top(
				container.PlaceWidget(dashboard.ProgressBar),
			),
			container.Bottom(
				container.Border(linestyle.Light),
				container.BorderTitle("Logs"),
				container.PlaceWidget(dashboard.Logs),
			),
		),
	)
	if err != nil {
		return nil, err
	}

	quitter := func(k *terminalapi.Keyboard) {
		if k.Key == 'q' || k.Key == 'Q' {
			term.Close()
		}
	}

	go dashboard.SetProgress(playbackProgress)
	go dashboard.WriteLogMessage(logs)

	if err := termdash.Run(ctx, term, c, termdash.KeyboardSubscriber(quitter)); err != nil {
		return nil, fmt.Errorf("Error running dashboard, err: %s", err.Error())
	}

	return dashboard, nil
}

// SetProgress gets playback progress by reading from the given channel and
// updating the dashboard progress bar accordingly.
func (d *Dashboard) SetProgress(playbackProgress chan PlaybackProgress) {
	for progress := range playbackProgress {
		var pct float64
		// dont divide by 0
		if progress.Current > 0 && progress.Total > 0 {
			pct = (float64(progress.Current) / float64(progress.Total)) * 100.0
		}
		if err := d.ProgressBar.Percent(int(pct)); err != nil {
			Errorlog.Errorf("Progress: %s, err: %v", progress.String(), err)
		}
	}
}

// WriteLogMessage gets log messages (mostly regarding polly processing status)
// by reading from the given channel and updating the dashboard text box accordingly.
func (d *Dashboard) WriteLogMessage(logs chan string) {
	for logMessage := range logs {
		if err := d.Logs.Write(logMessage); err != nil {
			Errorlog.Errorf("Log: %s, err: %v", logMessage, err)
		}
	}
}


main.go
package main

import (
	"bytes"
	"context"
	"flag"
	"fmt"
	"io"
	"os"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/polly"
	"github.com/aws/aws-sdk-go-v2/service/polly/types"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	log "github.com/sirupsen/logrus"
	"go.szostok.io/version"
	"go.szostok.io/version/printer"
)

// PlaybackProgress represents how far we have gotten in playing the audio
type PlaybackProgress struct {
	Total   int
	Current int
}

func (p *PlaybackProgress) String() string {
	return fmt.Sprintf("Total: %d, Current: %d", p.Total, p.Current)
}

const MAX_CHAR_COUNT = 200_000  // this is a polly limit per job
const DEFAULT_VOICE = "Matthew" // this can be overridden with cli flags

func main() {
	var ctx, cancel = context.WithCancel(context.Background())

	log.SetFormatter(&log.TextFormatter{
		FullTimestamp:   true,
		TimestampFormat: "2006-01-02 15:04:05",
	})

	// get user opts
	var s3Bucket string
	var awsProfile string
	var awsRegion string
	var voiceID string
	var inputFile string
	var outputFile string
	var dashboard bool
	var v bool
	flag.StringVar(&s3Bucket, "bucket", "", "s3 bucket to put the mp3 files")
	flag.StringVar(&awsProfile, "profile", "default", "aws profile to use")
	flag.StringVar(&awsRegion, "region", "us-west-2", "aws region to use")
	flag.StringVar(&voiceID, "voice", "Matthew", "voice to use")
	flag.StringVar(&inputFile, "input", "", "path the input text file, if this is specified STDIN will be ignored")
	flag.StringVar(&outputFile, "output", "output.mp3", "path the save the mp3, this will NOT play the audio")
	flag.BoolVar(&dashboard, "dashboard", false, "use a terminal dashboard")
	flag.BoolVar(&v, "version", false, "print version")
	flag.BoolVar(&v, "v", false, "print version")

	flag.Parse()

	if v {
		var verPrinter = printer.New()
		var info = version.Get()
		if err := verPrinter.PrintInfo(os.Stdout, info); err != nil {
			log.Fatal(err)
		}
		os.Exit(0)
	}

	// validate input
	if strings.TrimSpace(s3Bucket) == "" {
		log.Fatal("s3 bucket not spcecified")
	}

	if voiceID != DEFAULT_VOICE {
		var voices = types.VoiceId("").Values()
		var found bool
		for _, voice := range voices {
			if voice == types.VoiceId(voiceID) {
				found = true
				break
			}
		}
		if !found {
			log.Fatalf("VoiceID: %s is not an AWS Polly VoiceID", voiceID)
		}
	}

	var text string
	var err error
	if strings.TrimSpace(inputFile) != "" {
		b, err := os.ReadFile(strings.TrimSpace(inputFile))
		if err != nil {
			log.Fatalf("cannot read input file %s: %v", inputFile, err)
		}
		text = string(b)

	} else {
		text, err = readInput(os.Stdin)
		if err != nil {
			log.Fatalf("cannot read input %v", err)
		}
	}
	if text == "" {
		return
	}

	// create aws clients
	awsConfig, err := config.LoadDefaultConfig(ctx, config.WithSharedConfigProfile(awsProfile), config.WithRegion(awsRegion))
	if err != nil {
		log.Fatalf("failed to load SDK configuration, %v", err)
	}
	var pollyClient = polly.NewFromConfig(awsConfig)
	var s3Client = s3.NewFromConfig(awsConfig)
	var audioChan = make(chan *s3.GetObjectOutput, 5)
	var errors = make(chan error)
	var playbackProgress = make(chan PlaybackProgress)
	var logs = make(chan string)

	if !dashboard {
		go logOutput(playbackProgress, logs)
	}

	go playWithProgressBar(audioChan, playbackProgress, errors)
	go func() {
		if err := handleOutput(ctx, pollyClient, s3Client, audioChan, logs, s3Bucket, voiceID, text, outputFile); err != nil {
			log.Fatalf("handleOutput error: %s", err.Error())
		}
	}()

	if !dashboard {
		if err := <-errors; err != nil {
			fmt.Println(err)
		}
		cancel()
	} else {
		// this is a bit backwards because idealy NewDashboard would be run in a go routine above playWithProgressBar
		// however the term.Run() call cannot be done in a goroutine so it must be last in this main().
		go func() {
			if err := <-errors; err != nil {
				fmt.Println(err)
			}
			cancel()
		}()
		if _, err := NewDashboard(ctx, cancel, playbackProgress, logs); err != nil {
			log.Fatalf("failed to create dashboard, %v", err)
		}
	}
}

// handleOutput synthesizes text and either writes the result to a file or a channel for playing. File writing and playing are exclusize and is determined by cli flags.
func handleOutput(ctx context.Context, pollyClient *polly.Client, s3Client *s3.Client, audioChan chan *s3.GetObjectOutput, logs chan string, s3Bucket, voiceID, text, outputFile string) error {

	// splitting the input allows us to handle input that is larger than the max input size of polly (200k)
	var textSections = splitInput(text)
	logs <- fmt.Sprintf("The input text has been slpit into %d sections in order to comply with polly limits. \n", len(textSections))

	for _, section := range textSections {
		voice, s3File, err := synthesizeText(ctx, pollyClient, s3Client, logs, s3Bucket, voiceID, section)
		if err != nil {
			return fmt.Errorf("error from synthesisText: %v", err)
		}

		// output switch
		if strings.TrimSpace(outputFile) != "output.mp3" {
			body, err := io.ReadAll(voice.Body)
			if err != nil {
				return fmt.Errorf("error reading voice.Body: %v", err)
			}
			//nolint:gosec
			if err := os.WriteFile(outputFile, body, 0775); err != nil {
				return fmt.Errorf("error writing file %v", err)
			}
		} else {
			audioChan <- voice
		}

		// clean up s3 bucket
		if err := deleteS3File(ctx, s3Client, s3Bucket, s3File); err != nil {
			return fmt.Errorf("error deleting s3 files %v", err)
		}
	}
	close(audioChan)
	close(logs)
	return nil
}

// playWithProgressBar manages the progess bar and plays the audio
func playWithProgressBar(audioChan chan *s3.GetObjectOutput, playbackProgress chan PlaybackProgress, errors chan error) {

	for voice := range audioChan {

		// get the audio length using ffmpeg because polly doesnt return it
		var tempFile = "ffmpeg-detect-length-temp-file.mp3"
		ffmpegSound, err := io.ReadAll(voice.Body)
		if err != nil {
			errors <- fmt.Errorf("error reading voice.Body: %v", err)
			close(errors)
			return
		}

		voice.Body = io.NopCloser(bytes.NewBuffer(ffmpegSound))
		//nolint:gosec
		if err := os.WriteFile(tempFile, ffmpegSound, 0775); err != nil {
			errors <- fmt.Errorf("error writing file %v", err)
			close(errors)
			return
		}

		audioLength, err := getDuration(tempFile)
		if err != nil {
			errors <- fmt.Errorf("error getting length from ffmpeg: %v", err)
			close(errors)
			return
		}

		if err := os.RemoveAll(tempFile); err != nil {
			errors <- fmt.Errorf("error deleting temp file %v", err)
			close(errors)
			return
		}

		// get the progress bar going
		go func() {
			for i := 0; i < audioLength; i++ {
				playbackProgress <- PlaybackProgress{
					Current: i,
					Total:   audioLength,
				}
				time.Sleep(time.Second)
			}
		}()

		if err := play(voice.Body); err != nil {
			log.Fatalf("error playing audio %v", err)
		}
	}
	close(errors)
	close(playbackProgress)
}

// logOutput is used to print logs if the dashboard is not in use.
func logOutput(playbackProgress chan PlaybackProgress, logs chan string) {
	for playbackProgress != nil {
		select {
		case progress, ok := <-playbackProgress:
			if !ok {
				playbackProgress = nil
				continue
			}
			var pct float64
			// dont divide by 0
			if progress.Current > 0 && progress.Total > 0 {
				pct = (float64(progress.Current) / float64(progress.Total)) * 100.0
			}
			fmt.Printf("Progress: %.2f%% \n", pct)
		case log, ok := <-logs:
			if !ok {
				logs = nil
				continue
			}
			fmt.Printf("log: %s", log)
		}
	}
}
